runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineResources.java

```diff
@@ -40,34 +43,28 @@
 public class PipelineResources {
 
   /**
-   * Attempts to detect all the resources the class loader has access to. This does not recurse to
-   * class loader parents stopping it from pulling in resources from the system class loader.
+   * Detects all URLs that are present in all class loaders in between context class loader of
+   * calling thread and class loader of class passed as parameter. It doesn't follow parents above
+   * this class loader stopping it from pulling in resources from the system class loader.
    *
-   * @param classLoader The URLClassLoader to use to detect resources to stage.
-   * @throws IllegalArgumentException If either the class loader is not a URLClassLoader or one of
-   *     the resources the class loader exposes is not a file resource.
+   * @param cls Class whose class loader stops recursion into parent loaders
+   * @throws IllegalArgumentException no classloader in context hierarchy is URLClassloader or if
+   *     one of the resources any class loader exposes is not a file resource.
    * @return A list of absolute paths to the resources the class loader uses.
    */
-  public static List<String> detectClassPathResourcesToStage(ClassLoader classLoader) {
-    if (!(classLoader instanceof URLClassLoader)) {
-      String message =
-          String.format(
-              "Unable to use ClassLoader to detect classpath elements. "
-                  + "Current ClassLoader is %s, only URLClassLoaders are supported.",
-              classLoader);
-      throw new IllegalArgumentException(message);
+  public static List<String> detectClassPathResourcesToStage(Class<?> cls) {
+    Set<String> files = new HashSet<>();
+    ClassLoader stoppingLoader = cls.getClassLoader();
+    ClassLoader currentLoader = ReflectHelpers.findClassLoader();
+    while (currentLoader != null && stoppingLoader != currentLoader) {
+      files.addAll(extractResourcesFromClassLoader(currentLoader));
+      currentLoader = currentLoader.getParent();
     }
-
-    List<String> files = new ArrayList<>();
-    for (URL url : ((URLClassLoader) classLoader).getURLs()) {
-      try {
-        files.add(new File(url.toURI()).getAbsolutePath());
-      } catch (IllegalArgumentException | URISyntaxException e) {
-        String message = String.format("Unable to convert url (%s) to file.", url);
-        throw new IllegalArgumentException(message, e);
-      }
+    files.addAll(extractResourcesFromClassLoader(currentLoader));
+    if (files.isEmpty()) {
+      throw new IllegalArgumentException("Unable to use ClassLoader to detect classpath elements.");
```

Comment:
Why do we throw an error here? It doesn't seem consistent because we skip over non UrlClassloader below which may also contain required files.

Reply:
Actually, there is a test in `PipelineResourcesTest`, that explicitly tests this behavior. And I think it is fine, because otherwise we would again introduce breaking change for JDK >= 9, because we would just silently skip all classloaders and users would get weird exceptions at pipeline runtime (although it have worked on JDK8). I think it is better to let users know about that in a way that is more understandable.